lib/utils/compose.coffee

```diff
@@ -0,0 +1,732 @@
+
+# utilities
+
+renderProgressBar = (percentage, stepCount) ->
+	_ = require('lodash')
+	percentage = _.clamp(percentage, 0, 100)
+	barCount = stepCount * percentage // 100
+	spaceCount = stepCount - barCount
+	bar = "[#{_.repeat('=', barCount)}>#{_.repeat(' ', spaceCount)}]"
+	return "#{bar} #{_.padStart(percentage, 3)}%"
+
+pushProgressRenderer = (tty, prefix) ->
+	fn = (e) ->
+		{ error, percentage } = e
+		throw new Error(error) if error?
+		bar = renderProgressBar(percentage, 40)
+		tty.replaceLine("#{prefix}#{bar}\r")
+	fn.end = ->
+		tty.clearLine()
+	return fn
+
+buildLogCapture = (objectMode, buffer) ->
+	_ = require('lodash')
+	through = require('through2')
+
+	through { objectMode }, (data, enc, cb) ->
+		return cb(null, data) if not data?
+
+		# data from pull stream
+		if data.error
+			buffer.push("#{data.error}")
+		else if data.progress and data.status
+			buffer.push("#{data.progress}% #{data.status}")
+		else if data.status
+			buffer.push("#{data.status}")
+
+		# data from build stream
+		else
+			# normalise build log output here. it is somewhat ugly
+			# that this supposedly "passthrough" stream mutates the
+			# values before forwarding them, but it's convenient
+			# as it allows to both forward and save normalised logs
+
+			# convert to string, split to lines, trim each one and
+			# filter out empty ones.
+			lines = _(data.toString('utf-8').split(/\r?\n$/))
+				.map(_.trimEnd)
+				.reject(_.isEmpty)
+
+			# forward each line separately
+			lines.forEach (line) =>
+				buffer.push(line)
+				@push(line)
+
+			return cb()
+
+		cb(null, data)
+
+buildProgressAdapter = (inline) ->
+	through = require('through2')
+
+	stepRegex = /^\s*Step\s+(\d+)\/(\d+)\s*:\s+(.+)$/
+
+	[ step, numSteps, progress ] = [ null, null, undefined ]
+
+	through { objectMode: true }, (str, enc, cb) ->
+		return cb(null, str) if not str?
+
+		if inline
+			return cb(null, { status: str })
+
+		if /^Successfully tagged /.test(str)
+			progress = undefined
+		else
+			if (match = stepRegex.exec(str))
+				step = match[1]
+				numSteps = match[2]
+			if step?
+				str = "Step #{step}/#{numSteps}: #{str}"
+				progress = parseInt(step, 10) * 100 // parseInt(numSteps, 10)
+
+		cb(null, { status: str, progress })
+
+pullProgressAdapter = (outStream) ->
+	return (e) ->
+		{ status, id, percentage, error, errorDetail } = e
+		if status?
+			status = status.replace(/^Status: /, '')
+		if id?
+			status = "#{id}: #{status}"
+		if percentage is 100
+			percentage = undefined
+		outStream.write({
+			status,
+			progress: percentage,
+			error: errorDetail?.message ? error
+		})
+
+createSpinner = ->
+	chars = '|/-\\'
+	index = 0
+	-> chars[(index++) % chars.length]
+
+class BuildProgressUI
+	constructor: (tty, descriptors) ->
+		_ = require('lodash')
+		chalk = require('chalk')
+		through = require('through2')
+
+		services = _.map(descriptors, 'serviceName')
+		prefix = chalk.blue('[Build]') + '   '
```

Comment:
Why do we need to do manual prefixing etc etc here, instead of using/extending the existing logger that already handles that?

Reply:
Good point, I don't really remember the reason now, will explore the code and add a comment.

Reply:
Right, so the reason is `logger.coffee` magically prefixes the log line with `[Build]` etc., but it doesn't work well with a spinner we're also showing because the whole line is replaced on each update.